### PR TITLE
overlord/snapstate: make sure that snapd current symlink is not removed during refresh

### PR DESCRIPTION
**Pull request by bboozzoo**
When snap is refreshed, the current symlink would briefly go away between
unlink-current-snap and link-snap tasks. If the system gets rebooted at this
time, during startup all services from snaps will fail to start. This is caused
by the fact that /usr/bin/snap, which on Core is linked to
/snap/snapd/current/usr/bin/snap will be dangling, as /snap/snapd/current was
removed.

Address the problem by not removing current of the snapd snap, unless unlinking
is called during removal for the first install of the snapd snap on core or a
general snapd snap removal. In the snapd on core scenario, the /usr/bin/snap
will be invoked from the core snap anyway.

Supersedes #10066

### CODE REVIEW

**Comment by xnox** [comment]
No idea what this does, but just wanted to mention that it should be possible to update the target of symlink atomically =)
I.e. change where the symlink points to, without removing it first. Whilst both old and new locations exist.

File: overlord/snapstate/handlers.go [hunk 1]
```diff
@@ -836,16 +838,21 @@ func (m *SnapManager) doUnlinkCurrentSnap(t *state.Task, _ *tomb.Tomb) error {
...
-		// This task is only used for unlinking a snap during refreshes so we
-		// can safely hard-code this condition here.
-		RunInhibitHint: runinhibit.HintInhibitedForRefresh,
-	}
-	err = m.backend.UnlinkSnap(oldInfo, linkCtx, NewTaskProgressAdapterLocked(t))
-	if err != nil {
-		return err
+	if oldInfo.Type() != snap.TypeSnapd {
+		// snapd current symlink on the refresh path can only replaced
+		// by a symlink to a new revision of the snapd snap
```
Comment (anonymouse64) [inline overlord/snapstate/handlers.go:843]: this comment being inside the if was initially confusing to me and made me think that the `!= snap.TypeSnapd` was meant to be a `== snap.TypeSnapd`, but I see what the intention was, can you move this comment outside the if so it is more clear that this if branch is only taken in the case that we aren't the unlinking snapd snap?

i.e.

```suggestion
	// snapd current symlink on the refresh path can only replaced
	// by a symlink to a new revision of the snapd snap, so only do
	// the actual unlink if we're not working on the snapd snap
	if oldInfo.Type() != snap.TypeSnapd {
```

File: overlord/snapstate/handlers_link_test.go [hunk 2]
```diff
@@ -532,6 +532,50 @@ func (s *linkSnapSuite) TestDoUndoUnlinkCurrentSnapWithVitalityScore(c *C) {
...
+
+	// And observe the results.
+	var snapst snapstate.SnapState
+	err := snapstate.Get(s.state, "snapd", &snapst)
+	c.Assert(err, IsNil)
+	c.Check(snapst.Active, Equals, false)
+	c.Check(snapst.Sequence, HasLen, 1)
+	c.Check(snapst.Current, Equals, snap.R(20))
+	c.Check(task.Status(), Equals, state.DoneStatus)
+	// backend was not called to unlink the snap
```
Comment (anonymouse64) [inline overlord/snapstate/handlers_link_test.go:571]:
```suggestion
	// backend was called to unlink the snap
```

File: overlord/snapstate/handlers_link_test.go [hunk 3]
```diff
@@ -532,6 +532,50 @@ func (s *linkSnapSuite) TestDoUndoUnlinkCurrentSnapWithVitalityScore(c *C) {
 	c.Check(s.fakeBackend.ops, DeepEquals, expected)
 }
 
+func (s *linkSnapSuite) TestDoUnlinkSnapdUnlinks(c *C) {
```
Comment (anonymouse64) [inline overlord/snapstate/handlers_link_test.go:535]: I'm surprised we don't need to modify more tests, I would have expected us to have a test that is essentially `TestDoLinkSnapdNthInstall` which we would see now has a different set of backend operations (where now that test would not unlink the old snap and would just link the new snap). Can we add such a test here to be more clear about the new behavior?

**Review by anonymouse64 — COMMENTED** [review]
looks good to me, I think this makes sense, but I think we need more test coverage here

File: overlord/snapstate/handlers.go [hunk 4]
```diff
@@ -836,16 +838,21 @@ func (m *SnapManager) doUnlinkCurrentSnap(t *state.Task, _ *tomb.Tomb) error {
...
-	linkCtx := backend.LinkContext{
-		FirstInstall: false,
-		// This task is only used for unlinking a snap during refreshes so we
-		// can safely hard-code this condition here.
-		RunInhibitHint: runinhibit.HintInhibitedForRefresh,
-	}
-	err = m.backend.UnlinkSnap(oldInfo, linkCtx, NewTaskProgressAdapterLocked(t))
-	if err != nil {
-		return err
+	if oldInfo.Type() != snap.TypeSnapd {
```
Comment (pedronis) [inline overlord/snapstate/handlers.go:841]: do we test the skipping of this, I don't see a new test that mentions unlink-current-snap

**Review by pedronis — COMMENTED** [review]
code lgtm, question about testing

File: overlord/snapstate/handlers.go (hunk 4 above)
Comment (bboozzoo) [reply overlord/snapstate/handlers.go:841]: Added now.

File: overlord/snapstate/handlers_link_test.go (hunk 3 above)
Comment (bboozzoo) [reply overlord/snapstate/handlers_link_test.go:535]: I've added a snapstate level unit test.

**Review by pedronis — APPROVED** [review]
thank you

File: overlord/snapstate/snapstate_test.go [hunk 5]
```diff
@@ -6451,3 +6451,132 @@ func (s *snapmgrTestSuite) TestInstallModeDisableFreshInstallEnabledByHook(c *C)
 	c.Assert(op, Not(IsNil))
 	c.Check(op.disabledServices, HasLen, 0)
 }
+
+func (s *snapmgrTestSuite) TestSnapdRefreshTasks(c *C) {
```
Comment (anonymouse64) [inline overlord/snapstate/snapstate_test.go:6455]: thanks for this test

**Review by anonymouse64 — APPROVED** [review]
thanks for the additional tests here, lgtm

**Comment by bboozzoo** [comment]
@mvo5 `tests/main/snap-advise-command` is the only failure on ubunt-* here, I think we should land this PR.

**Pull request merged**